### PR TITLE
Add docker support

### LINKED ISSUE #24: Add Docker Support
Spinning this off a docker image will ease a lot of setup and deployment efforts. I don't have to install node and npm stuff on my host machine. Would help hugely with teams when it comes to development collaboration. Alas, but it's working on my machine!

### PR DESCRIPTION
**Pull request by eyedol**
**- Summary**
- Adds a script in the scripts folder that creates a docker machine, builds a docker image and runs the docker machine.
- Adds a folder called docker which contains docker template files.
- Updates `README` with clear instructions on how to run the script.

It fixes #24 

**- Test plan**
Run command below to create a default docker machine. Note, script is already executable.
```bash
./scripts/create-docker-machine-and-run-it
```

**- Description for the changelog**
- Added docker support for development.

**- A picture of a cute animal (not mandatory but encouraged)**

Girrafe

### CODE REVIEW

File: docker/Dockerfile.template [hunk 1]
```diff
@@ -0,0 +1,26 @@
+FROM node:6.10.0
+# Update packages
+RUN apt-get update && apt-get upgrade -y
+# Install dependencies
+RUN apt-get install build-essential wget -y -f
+# Set environment variable
+ARG RUN_AS=web
+ARG HUGO_VERSION=0.20.2
+ARG HUGO="hugo_${HUGO_VERSION}-64bit.deb"
+RUN wget "https://github.com/spf13/hugo/releases/download/v$HUGO_VERSION/$HUGO"
```
Comment (jhabdas) [inline docker/Dockerfile.template]: Will the deb package run under Ubuntu?

File: docker/Dockerfile.template [hunk 2]
```diff
@@ -0,0 +1,26 @@
+FROM node:6.10.0
```
Comment (jhabdas) [inline docker/Dockerfile.template]: Thoughts on using a micro container with a latest stable node?

File: docker/Dockerfile.template [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+# Set environment variable
+ARG RUN_AS=web
+ARG HUGO_VERSION=0.20.2
+ARG HUGO="hugo_${HUGO_VERSION}-64bit.deb"
+RUN wget "https://github.com/spf13/hugo/releases/download/v$HUGO_VERSION/$HUGO"
+RUN dpkg -i $HUGO
+RUN npm install --global gulp-cli
+# Setup user to run the app
+RUN useradd --user-group --create-home --shell /bin/false $RUN_AS
+ENV HOME=/home/$RUN_AS
```
Comment (jhabdas) [inline docker/Dockerfile.template]: Should we use ~ here?

File: scripts/create-docker-machine-and-run-it [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+#
+# Setup a docker machine for development
+set -eu
+
+readonly PROGNAME=$(basename "$0")
+readonly ARGS=("$@")
+# Docker settings
+MACHINE_NAME="devel"
+HUGO_VERSION=0.20.2
+NODE_VERSION=6.10.0
```
Comment (jhabdas) [inline scripts/create-docker-machine-and-run-it]: Any way to DRY this up with container version?

**Comment by eyedol** [comment]
@jhabdas Nope but I wanted to maintain consistency. Indentation with `EOF` has to be tabs not spaces though.

File: docker/Dockerfile.template (hunk 3 above)
Comment (jhabdas) [inline docker/Dockerfile.template]: Disregard.

File: scripts/create-docker-machine-and-run-it (hunk 4 above)
Comment (eyedol) [inline scripts/create-docker-machine-and-run-it]: Could you be a bit clearer? Do you mean put it in a function?

File: docker/Dockerfile.template (hunk 1 above)
Comment (eyedol) [inline docker/Dockerfile.template]: No. It's based off of debian:jessie

File: docker/Dockerfile.template (hunk 2 above)
Comment (eyedol) [inline docker/Dockerfile.template]: We could switch to `node:7.9.0-alpine`. It has smaller image footprint. The only issue is more work in installing hugo. I haven't played much with alpine based images.

File: docker/Dockerfile.template [hunk 5]
```diff
@@ -0,0 +1,26 @@
+FROM node:7.9.0
```
Comment (jhabdas) [inline docker/Dockerfile.template]: Guessing this is about ~500-600MB. Would prefer to see a Node microcontainer. Would likely bring the container size down about half a gig. https://www.iron.io/microcontainers-tiny-portable-containers/

File: docker/Dockerfile.template (hunk 1 above)
Comment (jhabdas) [inline docker/Dockerfile.template]: That could be helpful for running on RPi. Okay.

File: scripts/create-docker-machine-and-run-it (hunk 4 above)
Comment (jhabdas) [inline scripts/create-docker-machine-and-run-it]: > Could you be a bit clearer? Do you mean I put it in a function?

Container version using a fixed node version. This is also a fixed node version. Suggested approach is to create a config file to manage versions of both Hugo and container.

Comment (eyedol) [inline scripts/create-docker-machine-and-run-it]: The only place necessary to edit the version numbers is this file. I think it's okay. Some of the entries from the template files are overwritten when the script runs. The node and hugo versions always get overwritten.

File: docker/Dockerfile.template (hunk 5 above)
Comment (eyedol) [inline docker/Dockerfile.template]: Switched to `node:7.9.0-alpine`

**Comment by jhabdas** [comment]
Force push wiped discussion. Rats. Not sure where we left off. But I saw the alpine addition, and that's great. Does Alpine have a "latest" tag for their microcontainers? What's the size diff now between now and before?

**Comment by eyedol** [comment]
If you expand `Show outdated` you should see the previous chats. Yes, Alpine has a latest tag. I prefer using a specific version number than the latest tag so when something breaks you know which node version is breaking. Not sure about the diff size. But the changes I made aren't significant to affect the diff size.

**Comment by jhabdas** [comment]
Thanks. I will bow out now. Not a maintainer

**Comment by calavera** [comment]
From my point of view, this adds too much complexity for very little benefit.

At the end of the day, you're replacing having NPM and Hugo installed for having the Docker Engine, Docker Machine and Docker Compose installed. I rather ask users to install things that they directly need when working with this boilerplate.

**Comment by eyedol** [comment]
I can understand where you coming from but I was looking at it more from the point of view of users already having docker and docker compose installed. The main motivations for this were, I didn't want to install `npm` and `hugo` on my machine and more so as raised in #24

**Comment by calavera** [comment]
I appreciate it, but given that this is only a boilerplate, I rather not add all this. Since this repo is designed to be a minimal scaffolding, you can always add those scripts to your project.

By keeping it as simple as possible we don't tie it to any version of Hugo, this change would force a specific version, and we keep the dependencies to the minimum. Adding this would only create more confusion.

**Comment by eyedol** [comment]
No problem. Thanks for sharing your thoughts on it.

**Comment by jhabdas** [comment]
@eyedol The Jekyll gods [use Docker](https://github.com/jekyll/docker), and I have too along with static site generators to get [even the novice developer](https://habd.as/simple-websites-jekyll-docker/#comment-2013590199) up and running with a [The Twelve-Factor App](https://12factor.net).

In my experience developers who know how to debug things others do not (e.g. Node or Ruby) are prone to assuming if it's easy for them it must be easy for anyone, which is not the case.

As it stands, this Boilerplate uses both a Node environment and makes assumptions about Hugo support which are likely to lead to a cascade of deployment issues for individuals not containerizing their apps.

@eyedol You may want to float your idea past the Hugulp creator. And if he doesn't bite I'd be glad to work together to create a simple Docker setup we can use to enable others to build maintainable apps that scale and are easy to deploy on any operating system, including Jessie for the Raspberry Pi.

**Comment by eyedol** [comment]
@jhabdas Thanks for your comments and the mentioned of Hugulp. I have opened an [issue](https://github.com/jbrodriguez/hugulp/issues/23) to see if they will be open to having docker support. I like the collaboration idea. Let's see how this goes.

**Pull request closed without merging**